### DISCUSSION TITLE
SIGNUPS_ALLOWED not stopping new user creation

### DISCUSSION THREAD

**Discussion by JBFUK** [dani-garcia/vaultwarden#2030] [Q&A]
### Subject of the issue

SIGNUPS_ALLOWED not stopping new user creation

### Deployment environment

Latest Docker Image

* Install method: Docker Compose

* Clients used: Web-Vault

* Reverse proxy and version: Nginx

* MySQL/MariaDB or PostgreSQL version: MariaDB

* Other relevant details:

### Steps to reproduce
Set environment variable SIGNUPS_ALLOWED=false via docker-compose file

### Expected behaviour
Expect new user creation from the main login page to fail with an error (as described in WIKI)

### Actual behaviour
Allows new user creation and subsequent login

**Comment by jjlin** [comment]
You probably have `SIGNUPS_DOMAINS_WHITELIST` set as well.

See https://github.com/dani-garcia/vaultwarden/wiki/Disable-registration-of-new-users#restricting-registrations-to-certain-email-domains.

**Comment by pepa65** [comment]
So we have SIGNUPS_ALLOWED [S], INVITATIONS_ALLOWED [I], and SIGNUPS_DOMAINS_WHITELIST [W] all interacting with each other:

S I W Result
0 0 0 Self-signup and Invititations impossible
0 0 1 Self-signup and Invititations impossible
0 1 0 Invititations possible, unlimited
0 1 1 Invititations possible, limited by the Whitelist
1 0 0 Self-signup possible, unlimited
1 0 1 Self-signup possible, limited by the Whitelist
1 1 0 Self-signup and Invititations possible, unlimited 
1 1 1 Self-signup and Invititations possible, limited by the Whitelist

The above is what makes logical sense. But you're saying that currently when SIGNUPS_DOMAINS_WHITELIST is set, it is like "Self-signup and Invititations possible, limited by the Whitelist"? If so, wouldn't it make sense to see that as a bug, and fix it??

**Comment by jjlin** [comment]
These settings were implemented by different people at different times, so the design is not always the most cohesive.

`SIGNUPS_DOMAINS_WHITELIST` is documented as

https://github.com/dani-garcia/vaultwarden/blob/9930a0d75224202dbc355e945724aa545bedebe2/.env.template#L179-L181

The original implementation required `SIGNUPS_ALLOWED=false` to be set. Since the default is `SIGNUPS_ALLOWED=true`, a quite common issue was to set only `SIGNUPS_DOMAINS_WHITELIST` and expect that to limit the signups, which wasn't how it worked. So later `SIGNUPS_DOMAINS_WHITELIST` was changed to simply override `SIGNUPS_ALLOWED`, because why would you set `SIGNUPS_DOMAINS_WHITELIST` if you didn't expect that to control signups?

With this, you could think of it either as "signups allowed, but only from these domains" or "signups not allowed, except from these domains", both of which seem like logical interpretations to me.

If you're interested in proposing and implementing another design that is still backward compatible, feel free.

**Comment by pepa65** [comment]
Backward compatible meaning: we keep the same variables and the current results need to stay as they are? And then introduce new variables to attain the same range of outcomes more sensibly??

I think the outcomes I listed are fairly useful, and with the 3 boolean variables, there are only 2 outcomes that are overlapping, ie. if SIGNUPS_ALLOWED [S] and INVITATIONS_ALLOWED are both false, the value of SIGNUPS_DOMAINS_WHITELIST is irrelevant, and that makes sense.

@jjlin My proposal would be along the lines of the list of outcomes in my previous post. Do you think that is workable?

**Comment by jjlin** [comment]
I would say backward compatible means something that is expected to remain working for the vast majority of existing deployments.

Given the existing constraints and the fact that people can achieve the desired results if they read the docs, I'm not sure polishing this a lot is worth the effort, and it's not work I'm personally interested in. But other people who feel strongly about it are certainly welcome to step up.

**Comment by pepa65** [comment]
I guess the bug reporter here would need to set all 3 to FALSE to achieve his desired outcome.

I all desired outcomes are already achievable, we can let this rest, until someone want to achieve something that the current setup does not cater to. Example: wanting to allow invitations that are limited to the whitelist but no self-signup. Then it could be fixed, but warn people with existing installations that Vaultwarden got a bit more specific.

**Comment by BlackDex** [comment]
There is no reason to set a whitelist and not have it able to allow signups or invites for that domain. If that is what someone intends to achieve then they need to clear all and add users manually via the admin interface. Setting the whitelist is as @jjlin already mentioned is a kind of boolean flag for SIGNUPS_ALLOWED at the same time.

I am also not interested into trying to iron this. It will probably break, and i also don't like adding more and more variables, that would make it even worse i think.

**Reply by pepa65** [reply]
> There is no reason to set a whitelist and not have it able to allow signups or invites for that domain.

S I W Result
0 0 0 Self-signup and Invititations impossible
0 0 1 Self-signup and Invititations impossible
0 1 0 Invititations possible, unlimited
0 1 1 Invititations possible, limited by the Whitelist
1 0 0 Self-signup possible, unlimited
1 0 1 Self-signup possible, limited by the Whitelist
1 1 0 Self-signup and Invititations possible, unlimited
1 1 1 Self-signup and Invititations possible, limited by the Whitelist

All these achievements/results/outcomes make sense, except that the first 2 have the same outcome. Right now, if you set the Whitelist, it also allows Self-signup (I'm guessing without setting the variable to make it explicit). Does it also automatically allow Invitations??

I don't think it is hard to make this work, and I am willing to look into this. These variables can just be checked whenever a Self-signup or Invitation occurs, and the whitelist can (or won't, depending on the variable) be taken into account.

**Comment by JBFUK** [comment]
This is all I have sent for the vaultwarden container:

```
 environment:
   - "DATABASE_URL=mysql://${MYSQL_USER}:${MYSQL_PASSWORD}@mariadb/${MYSQL_DATABASE}"
   - "ADMIN_TOKEN=${ADMIN_TOKEN}"
   - "RUST_BACKTRACE=1"
   - "SIGNUPS_ALLOWED=false"
```

What I want is to allow invitations and users to sign up IF they are invited (regardless of domain) but to prevent Joe Public who just happened to find his way to my vault login page from signing up uninvited.

My understanding from the WIKI was that setting "SIGNUPS_ALLOWED=false" would achieve this.

**Comment by BlackDex** [comment]
Did you do a `docker-compose up -d` after you modified it?
If not, these changes didn't had any effect.

Please go to `/admin/diagnostics` and generate a support string there, so that we can see the running config.

**Comment by JBFUK** [comment]
Yes I did. I actually did a `docker-compose down` followed by `docker-compose up -d`.

What I have noticed is that the https://myvault/admin/ interface under 'General Settings' shows signups allowed.  I'm just going to try disabling from there to see whether that works.  If so then it seems like the ENV variable is not being applied and/or overriding the admin setting?

[EDIT]

Confirmed, it works fine with signups disabled via the admin UI.  Do you still want a support string dump?

**Comment by BlackDex** [comment]
I think you actually need to change your docker-compose file.
```yml
environment:
  - TZ=UTC
  - DOMAIN=https://my.domain.tld
  - SIGNUPS_ALLOWED=false
  - ADMIN_TOKEN=${SECRET_ADMIN_TOKEN}
  - RUST_BACKTRACE=1
```

You have quotes `"` around the whole values, i think that breaks it in some way.
And i even think it is not using MySQL this way, but that I'm not sure.

**Comment by JBFUK** [comment]
The admin token takes effect and database is working fine.

I take the point about the quotes around the variables.  Usually I would not have these but in the interest of 'following the instructions' fully I included them - as this is what the vaultwarden wiki shows in its example docker-compose template[ [Here].](https://github.com/dani-garcia/vaultwarden/wiki/Using-the-MariaDB-%28MySQL%29-Backend)

**Reply by BlackDex** [reply]
I think it is time to check and verify if all the wiki doc's are still valid hehe.
This wiki doesn't use them: https://github.com/dani-garcia/vaultwarden/wiki/Using-Docker-Compose

**Comment by BlackDex** [comment]
@JBFUK, i just did a small test using docker-compose, and for me this works without any issues, with and without `"`.
What i think could have happened here, is that you may saved the config through the admin interface before, and changing settings via the `environment:` after that doesn't work for those variables. That only works for Read-Only configuration variables.

**Reply by JBFUK** [reply]
I think you're right.  I probably changed something else in the admin pages and once I'd hit 'save' any values there were stored in the database and take priority?  The question is then can I actually delete the config from the DB so that the env variables take effect?

**Comment by Glideh** [comment]
I can still Create Account although I explicitly disabled signup:

### My environment (Generated via diagnostics page)
* Vaultwarden version: v1.26.0
* Web-vault version: v2022.10.0
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** ADMIN_TOKEN


```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://***********.*****.***",
  "domain_origin": "*****://***********.*****.***",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 5 * * * *",
  "emergency_request_timeout_schedule": "0 5 * * * *",
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_hints_allowed": true,
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": null,
  "smtp_from": "***********@*****.***",
  "smtp_from_name": "Vaultwarden Gprod",
  "smtp_host": "*****.*******.***",
  "smtp_password": null,
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

Do you see something I have wrong ?

**Reply by BlackDex** [reply]
Can you just signup any mail address? Even some fake one like blaat@my.local? Or are those addresses invited via the admin or org? Because invites are allowed.

**Reply by Glideh** [reply]
Ok, I didn't actually try to create a user, well I can't, seems like the expected behavior.
I thought the signup button would disappear when we disable the signup.
Thanks for your quick answer anyway,

**Reply by Codelica** [reply]
This does seem rather odd...  

So with both `INVITATIONS_ALLOWED=false` and `SIGNUPS_ALLOWED=false` and `SIGNUPS_DOMAINS_WHITELIST` unset, the expected behavior is for the "Create account" link to still be offered, the `/#/register` page to still browseable, but any submitted form is just blocked?

**Reply by BlackDex** [reply]
This isn't odd. There web-vault is a static application which doesn't have flags to hide the register button. Also, Vaultwarden isn't able to on-the-fly modify this because those files are cached by there clients.

And.. if you invite someone via the admin interface, they still need to follow those steps! So the register page still needs to be accessable.

Remember that Vaultwarden only makes some small adjustments to the web-vault to make it work and keep compatible with Bitwarden.

**Reply by Codelica** [reply]
Thanks, that is definitely enlightening.  I had no idea there was that level of re-use with the web-vault.  Should have done some late-night FAQ reading first.  :)  Easy enough to remove the link via proxy filtering like NGINX's sub_filter.  Thanks.
